### ISSUE TITLE
Program

### ISSUE THREAD

**Issue by somaiah-ui** [somaiah-ui/Placement-Training-2026-Team-Project#2]
import torch
import torch.nn.functional as F
from sentence_transformers import SentenceTransformer
from transformers import AutoModelForCausalLM, AutoTokenizer

# 1. KNOWLEDGE BASE
KNOWLEDGE_BASE = [
    {
        "id": 1,
        "title": "Python List vs Tuple",
        "content": (
            "In Python, lists are mutable, meaning elements can be added,"
            " removed, or changed. They use square brackets []. Tuples are"
            " immutable, use parentheses (), and are faster with lower memory"
            " overhead."
        ),
    },
    {
        "id": 2,
        "title": "Self-Attention Mechanism",
        "content": (
            "Self-attention allows a Transformer model to relate different"
            " positions of a single sequence to compute a representation. It"
            " relies on Query (Q), Key (K), and Value (V) vectors with the"
            " formula: Softmax(QK^T / sqrt(d_k)) * V."
        ),
    },
    {
        "id": 3,
        "title": "Database Indexing & B-Trees",
        "content": (
            "Database indexes speed up data retrieval operations at the cost of"
            " slower writes. B-Trees and B+ Trees maintain balanced depth,"
            " guaranteeing O(log N) search, insertion, and deletion times."
        ),
    },
    {
        "id": 4,
        "title": "Transformer Positional Encodings",
        "content": (
            "Since Transformers lack recurrent loops and process all tokens"
            " simultaneously, positional encodings (sinusoidal functions or"
            " learned embeddings) are injected to provide token order"
            " information."
        ),
    },
]

# 2. TRANSFORMER ENCODER: Semantic Search
class TransformerRetriever:

  def __init__(self, model_name="all-MiniLM-L6-v2", device="cpu"):
    print(f"[Init] Loading Transformer Encoder ({model_name})...")
    self.device = device
    self.encoder = SentenceTransformer(model_name, device=device)
    self.doc_texts = [doc["content"] for doc in KNOWLEDGE_BASE]

    # Precompute vector representations
    embeddings = self.encoder.encode(self.doc_texts, convert_to_tensor=True)
    self.doc_embeddings = F.normalize(embeddings, p=2, dim=1)

  def search(self, query: str, top_k: int = 1):
    query_embedding = self.encoder.encode(
        query, convert_to_tensor=True, device=self.device
    )
    query_embedding = F.normalize(query_embedding, p=2, dim=0)

    # Cosine similarity calculation via dot product
    cosine_scores = torch.mv(self.doc_embeddings, query_embedding)
    top_scores, top_indices = torch.topk(cosine_scores, k=top_k)

    results = []
    for score, idx in zip(top_scores.tolist(), top_indices.tolist()):
      results.append({"doc": KNOWLEDGE_BASE[idx], "score": score})
    return results

# 3. LLM GENERATOR: Causal LM
class LLMGenerator:

  def __init__(
      self, model_id="Qwen/Qwen2.5-Coder-0.5B-Instruct", device="cuda"
  ):
    print(f"[Init] Loading Decoder LLM ({model_id}) on {device}...")
    self.device = device
    self.tokenizer = AutoTokenizer.from_pretrained(model_id)
    self.model = AutoModelForCausalLM.from_pretrained(
        model_id,
        torch_dtype=torch.float16 if device == "cuda" else torch.float32,
        device_map="auto" if device == "cuda" else None,
    )
    if device == "cpu":
      self.model = self.model.to("cpu")

  def answer_with_context(self, question: str, retrieved_context: str) -> str:
    system_instruction = (
        "You are a technical assistant. Use only the provided context to answer"
        " the user's question accurately. If the context does not contain the"
        " answer, state that you do not know."
    )

    prompt = (
        f"<|im_start|>system\n{system_instruction}<|im_end|>\n"
        f"<|im_start|>user\nContext:\n{retrieved_context}\n\nQuestion:"
        f" {question}<|im_end|>\n"
        "<|im_start|>assistant\n"
    )

    inputs = self.tokenizer(prompt, return_tensors="pt").to(self.device)

    with torch.no_grad():
      outputs = self.model.generate(
          **inputs,
          max_new_tokens=120,
          temperature=0.1,
          do_sample=False,
          pad_token_id=self.tokenizer.eos_token_id,
      )

    full_output = self.tokenizer.decode(outputs[0], skip_special_tokens=True)
    return full_output.split("assistant\n")[-1].strip()

# ==========================================
# 4. EXECUTION
# ==========================================
# Detect hardware accelerator
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Active Runtime Device: {device.upper()}")

retriever = TransformerRetriever(device=device)
llm = LLMGenerator(device=device)

# Query test
query = "How do transformers know the order of words without an RNN?"

# Retrieve context via Transformer embeddings
matches = retriever.search(query, top_k=1)
top_doc = matches[0]["doc"]

print("\n" + "=" * 45)
print(f"Question: {query}")
print(f"Retrieved Node: {top_doc['title']} (Score: {matches[0]['score']:.4f})")
print("-" * 45)

# Generate response with LLM
context_str = f"Title: {top_doc['title']}\nDetails: {top_doc['content']}"
response = llm.answer_with_context(
    question=query, retrieved_context=context_str
)

print(f"LLM Answer:\n{response}")
print("=" * 45)
